### PR TITLE
Implementing `--time-report`

### PR DESCRIPTION
**Pull request by Abdelrahman-Kh-Fouad**
About #344 
Now, we can see parsering time.
```bash
$ ./src/bin/lpython --parser-time a.py
Parsing time: 5.32848 ms
```
I think we can add afile for comparing some files, or add feature on `run_test.py` to make this.

### CODE REVIEW

**Comment by certik** [comment]
I think this is a good idea. Why not to simply add the timing information for the "--show-ast" code as an option if the `--time-report` option is passed? LFortran does it that way for some backends. So we can eventually add it to all stages of the compiler.

**Comment by Abdelrahman-Kh-Fouad** [comment]
> I think this is a good idea. Why not to simply add the timing information for the "--show-ast" code as an option if the `--time-report` option is passed? LFortran does it that way for some backends. So we can eventually add it to all stages of the compiler.

ok, and if comilation fall in any stage report will show times of stages before this point.

**Comment by Abdelrahman-Kh-Fouad** [comment]
@certik 

- i implemented `--time-report` like LFortran and it called from `run_benchmark.py` to find parsing time.
- i made `run_benchmark.py` to measure time of `benchmark` dir files.
  * in `benchmarks.toml` we can customize every file to add it any kind of comparison we want {parse or src -> ast, etc .. }  
  * and if want to add a comparison we should add class that represent times for files, now there is just parser comparison.
  * after comparison there is a graph of data will be printed, and it shows us the time of every file in millisecond like : 
   img

File: src/bin/lpython.cpp [hunk 1]
```diff
@@ -222,7 +221,7 @@ int emit_cpp(const std::string &infile,
     std::cout << res.result;
     return 0;
 }
-
+#define HAVE_LFORTRAN_LLVM 1
```
Comment (certik) [inline src/bin/lpython.cpp:224]: I think this should not be here, this is defined by cmake in `config.h`.

**Comment by certik** [comment]
Interesting idea to run benchmarks like that in the `benchmarks` subdirectory --- we should probably do that. But let's submit that as a separate PR. In this PR, let's just implement the `--time-report`.

File: src/bin/lpython.cpp [hunk 2]
```diff
@@ -269,55 +268,84 @@ int emit_llvm(const std::string &infile,
...
+    times.push_back(std::make_pair("Reading_file", std::chrono::duration<double, std::milli>(end_file - start_file).count()));
+
+    auto start_parser = std::chrono::high_resolution_clock::now();
+    LFortran::Result < LFortran::LPython::AST::ast_t * > r = parse_python_file(
+            al, runtime_library_dir, infile, diagnostics, compiler_options.new_parser);
+    auto end_parser = std::chrono::high_resolution_clock::now();
+    times.push_back(std::make_pair("parsing", std::chrono::duration<double, std::milli>(end_parser - start_parser).count()));
     std::cerr << diagnostics.render(input, lm, compiler_options);
     if (!r.ok) {
+        print_report(times, time_report);
```
Comment (certik) [inline src/bin/lpython.cpp:305]: I have to test it, it seems it would always print the report if the parsing fails?

File: src/bin/lpython.cpp [hunk 3]
```diff
@@ -269,55 +268,84 @@ int emit_llvm(const std::string &infile,
...
     // ASR -> LLVM
+    auto start_asr_to_llvm = std::chrono::high_resolution_clock::now();
     LFortran::PythonCompiler fe(compiler_options);
     LFortran::LLVMEvaluator e(compiler_options.target);
-    std::unique_ptr<LFortran::LLVMModule> m;
-    LFortran::Result<std::unique_ptr<LFortran::LLVMModule>>
-        res = fe.get_llvm3(*asr, diagnostics);
+    std::unique_ptr <LFortran::LLVMModule> m;
+    LFortran::Result <std::unique_ptr<LFortran::LLVMModule>>
+            res = fe.get_llvm3(*asr, diagnostics);
```
Comment (certik) [inline src/bin/lpython.cpp:333]: The above seem like just formatting changes. I would do those as either separate PR, or at least separate commits. I think it looks better if written as `std::unique_ptr<LFortran::LLVMModule>` rather than `std::unique_ptr <LFortran::LLVMModule>`, doesn't it?

File: src/bin/lpython.cpp (hunk 2 above)
Comment (Abdelrahman-Kh-Fouad) [reply src/bin/lpython.cpp:305]: yes, and even if ast to asr  or asr to llvm, it will print last stage even if it faild.
i thought it will be usefull to know time even if stage faild, but i  will remove it.

**Comment by Abdelrahman-Kh-Fouad** [comment]
> Interesting idea to run benchmarks like that in the `benchmarks` subdirectory --- we should probably do that. But let's submit that as a separate PR. In this PR, let's just implement the `--time-report`.

ok, i will make for `--show-time`.

File: src/bin/lpython.cpp (hunk 1 above)
Comment (Abdelrahman-Kh-Fouad) [reply src/bin/lpython.cpp:224]: yes, you're right, i just put it there just to test something and i forget to erasr it,sorry.

File: src/bin/lpython.cpp (hunk 3 above)
Comment (Abdelrahman-Kh-Fouad) [reply src/bin/lpython.cpp:333]: i made this formate by ide formatter, but i change it no problem.

Comment (certik) [reply src/bin/lpython.cpp]: Which formatter do you use? We should probably setup a `clang-format` config file for the project.

Comment (Abdelrahman-Kh-Fouad) [reply src/bin/lpython.cpp]: it just a formatter with clion ide with no standard i think, but i will add clang to be the formatter.

File: src/bin/lpython.cpp (hunk 2 above)
Comment (certik) [reply src/bin/lpython.cpp:305]: I think it would be useful even if it fails.

Comment (Abdelrahman-Kh-Fouad) [reply src/bin/lpython.cpp:305]: ok, thats fine i will bring it back.

File: src/bin/lpython.cpp [hunk 4]
```diff
@@ -269,55 +267,84 @@ int emit_llvm(const std::string &infile,
...
         const std::string &infile,
         const std::string &outfile,
         const std::string &runtime_library_dir,
-        CompilerOptions &compiler_options)
-{
-    Allocator al(4*1024);
+        const bool time_report,
+        CompilerOptions &compiler_options) {
+
+    Allocator al(4 * 1024);
```
Comment (certik) [inline src/bin/lpython.cpp:285]: If you want to make formatting changes, you can, but please send them as a separate PR. In a PR where you are implementing new functionality, please do not change other code by adding white spaces, as it makes it harder to review, and also it seems to me these formatting changes are not necessarily making it more readable, but we should have that discussion on a separate PR, so that we can focus here on the actual functionality.

**Comment by certik** [comment]
If you wouldn't mind extracting all pure formatting changes into a separate PR, I would really appreciate it. It I suggested a few above, but there are more.
